File: dotnet/Families/GameX.Gamebryo/Platforms/NifObjectBuilder.cs
using GameX.Gamebryo.Formats.Nif;
using OpenStack.Gfx;
using System;

namespace GameX.Gamebryo.Platforms;

public static class NifObjectBuilder {
    public const int YardInMWUnits = 64;
    public const float MeterInYards = 1.09361f;
    public const float MeterInUnits = MeterInYards * YardInMWUnits;
    public const int MarkerLayer = 0;
    public const bool KinematicRigidbody = false;

    public static MaterialStd2Prop ToMaterialProp(NiAVObject s) {
        // find relevant properties.
        NiTexturingProperty tex = null;
        NiMaterialProperty mat = null;
        NiAlphaProperty alpha = null;
        foreach (var t in s.Properties) {
            var prop = t.Value;
            if (prop is NiTexturingProperty tp) tex = tp;
            else if (prop is NiMaterialProperty mp2) mat = mp2;
            else if (prop is NiAlphaProperty ap) alpha = ap;
        }

        // create the material properties.
        var mp = new MaterialStd2Prop();

        // apply alphaProperty
        if (alpha != null) {
            /*
            14 bits used:
            1 bit for alpha blend bool
            4 bits for src blend mode
            4 bits for dest blend mode
            1 bit for alpha test bool
            3 bits for alpha test mode
            1 bit for zwrite bool ( opposite value )
            Bit 0 : alpha blending enable
            Bits 1-4 : source blend mode 
            Bits 5-8 : destination blend mode
            Bit 9 : alpha test enable
            Bit 10-12 : alpha test mode
            Bit 13 : no sorter flag ( disables triangle sorting ) ( Unity ZWrite )
            */
            var flags = (ushort)alpha.Flags;
            var srcbm = (byte)(BitConverter.GetBytes(flags >> 1)[0] & 15);
            var dstbm = (byte)(BitConverter.GetBytes(flags >> 5)[0] & 15);
            mp.ZWrite = BitConverter.GetBytes(flags >> 15)[0] == 1;
            mp.AlphaBlended = (flags & 0x01) != 0;
            mp.SrcBlendMode = (GfxBlendMode)Math.Min((int)srcbm, 10);
            mp.DstBlendMode = (GfxBlendMode)Math.Min((int)dstbm, 10);
            mp.AlphaTest = (flags & 0x100) != 0;
            mp.AlphaCutoff = (float)alpha.Threshold / 255;
        }

        // apply materialProperty
        if (mat != null) {
            mp.Alpha = mat.Alpha;
            mp.DiffuseColor = mat.DiffuseColor.AsColor;
            mp.EmissiveColor = mat.EmissiveColor.AsColor;
            mp.SpecularColor = mat.SpecularColor.AsColor;
            mp.Glossiness = mat.Glossiness;
        }

        // apply texturingProperty
        if (tex != null && tex.TextureCount > 0) {
            var mt = mp.Textures;
            if (tex.BaseTexture != null) mt.Add("Main", tex.BaseTexture.Source.Value.FileName);
            if (tex.DarkTexture != null) mt.Add("Dark", tex.DarkTexture.Source.Value.FileName);
            if (tex.DetailTexture != null) mt.Add("Detail", tex.DetailTexture.Source.Value.FileName);
            if (tex.GlossTexture != null) mt.Add("Gloss", tex.GlossTexture.Source.Value.FileName);
            if (tex.GlowTexture != null) mt.Add("Glow", tex.GlowTexture.Source.Value.FileName);
            if (tex.BumpMapTexture != null) mt.Add("Bump", tex.BumpMapTexture.Source.Value.FileName);
        }
        return mp;
    }

    public static bool IsMarkerFileName(string name) => name.ToLowerInvariant() switch {
        "marker_light" or "marker_north" or "marker_error" or "marker_arrow" or "editormarker" or "marker_creature" or "marker_travel" or "marker_temple" or "marker_prison" or "marker_radius" or "marker_divine" or "editormarker_box_01" => true,
        _ => false,
    };
}

File: python/gamex/families/Gamebryo/platforms/nifobjectbuilder.py
import sys, os
from numpy import ones, zeros
from openstk.gfx.gfx import GfxBlendMode, MaterialStd2Prop
from gamex.families.Gamebryo.formats.nif import NiAVObject, NiTexturingProperty, NiMaterialProperty, NiAlphaProperty

class NifObjectBuilder:
    YardInMWUnits: int = 64
    MeterInYards: float = 1.09361
    MeterInUnits: float = MeterInYards * YardInMWUnits;
    MarkerLayer: int = 0
    KinematicRigidbody: bool = False

    @staticmethod
    def toMaterialProp(s: NiAVObject) -> MaterialStd2Prop:
        # find relevant properties.
        tex: NiTexturingProperty = None
        mat: NiMaterialProperty = None
        alpha: NiAlphaProperty = None
        for t in s.properties:
            prop = t.value
            if isinstance(prop, NiTexturingProperty): tex = prop
            elif isinstance(prop, NiMaterialProperty): mat = prop
            elif isinstance(prop, NiAlphaProperty): alpha = prop

        # create the material properties.
        mp = MaterialStd2Prop()

        # apply alphaProperty
        if alpha:
            '''
            14 bits used:
            1 bit for alpha blend bool
            4 bits for src blend mode
            4 bits for dest blend mode
            1 bit for alpha test bool
            3 bits for alpha test mode
            1 bit for zwrite bool ( opposite value )
            Bit 0 : alpha blending enable
            Bits 1-4 : source blend mode 
            Bits 5-8 : destination blend mode
            Bit 9 : alpha test enable
            Bit 10-12 : alpha test mode
            Bit 13 : no sorter flag ( disables triangle sorting ) ( Unity ZWrite )
            '''
            flags = int(alpha.flags)
            srcbm = (flags >> 1).to_bytes(2, byteorder='little')[0] & 15
            dstbm = (flags >> 5).to_bytes(2, byteorder='little')[0] & 15
            mp.zwrite = (flags >> 15).to_bytes(2, byteorder='little')[0] == 1
            mp.alphaBlended = (flags & 0x01) != 0
            mp.srcBlendMode = GfxBlendMode(min(srcbm, 10))
            mp.dstBlendMode = GfxBlendMode(min(dstbm, 10))
            mp.alphaTest = (flags & 0x100) != 0
            mp.alphaCutoff = alpha.threshold / 255.0

        # apply materialProperty
        if mat:
            mp.alpha = mat.alpha
            mp.diffuseColor = mat.diffuseColor.asColor
            mp.emissiveColor = mat.emissiveColor.asColor
            mp.specularColor = mat.specularColor.asColor
            mp.glossiness = mat.glossiness

        # apply texturingProperty
        if tex and tex.textureCount > 0:
            mt = mp.textures
            if tex.gaseTexture: mt['Main'] = tex.baseTexture.source.value.fileName
            if tex.darkTexture: mt['Dark'] = tex.darkTexture.source.value.fileName
            if tex.detailTexture: mt['Detail'] = tex.detailTexture.source.value.fileName
            if tex.glossTexture: mt['Gloss'] = tex.glossTexture.source.value.fileName
            if tex.glowTexture: mt['Glow'] = tex.glowTexture.source.value.fileName
            if tex.bumpMapTexture: mt['Bump'] = tex.bumpMapTexture.source.value.fileName
        return mp

    @staticmethod
    def isMarkerFileName(self, name: str) -> bool:
        match name.lower():
            case 'marker_light' | 'marker_north' | 'marker_error' | 'marker_arrow' | 'editormarker' | 'marker_creature' | 'marker_travel' | 'marker_temple' | 'marker_prison' | 'marker_radius' | 'marker_divine' | 'editormarker_box_01': return True
            case _: return False
